crates/sui-types/src/temporary_store.rs

```diff
@@ -772,6 +789,55 @@ impl<S> TemporaryStore<S> {
     }
 }
 
+impl<S: GetModule> TemporaryStore<S> {
+    /// Check that this transaction neither creates nor destroys SUI. This should hold for all txes except
+    /// the epoch change tx, which mints staking rewards equal to the gas fees burned in the previous epoch.
+    /// This intended to be called *after* we have charged for gas + applied the storage rebate to the gas object,
+    /// but *before* we have updated object versions
+    pub fn check_sui_conserved(&self) {
+        let gas_summary = &self.gas_charged.as_ref().unwrap().2;
+        let storage_fund_rebate_inflow =
+            gas_summary.storage_fund_rebate_inflow(self.storage_rebate_rate);
+        // storage gas cost should be equal to total rebates of mutated objects + storage fund rebate inflow (see below)
+        assert_eq!(
+            gas_summary.storage_cost,
+            output_rebate_amount + storage_fund_rebate_inflow
+        );
+
+        // total SUI in input objects
+        let input_sui = self.mutable_input_refs.iter().fold(0, |acc, o| {
+            acc + self
+                .input_objects
+                .get(&o.0)
+                .unwrap()
+                .get_total_sui(&self.store)
+                .unwrap()
+        });
+        // TODO: if a dynamic field object O is written by this tx, count get_total_sui(pre_tx_value(O)) as part of input_sui
```

Comment:
Wondering if @amnn or @tnowacki have ideas here

Reply:
Bit confused about the nature of these checks as I'm not sure how you are going to account for wrapped/unwrapped coins. 

Concerns about wrapping aside, you could use `modified_at_versions` to load all of the objects (dynamic fields included) of the objects used in the transaction, and compare it against the writes?

Reply:
The new `object.get_total_suit` function I added in this PR handles wrapped coins--it recurses through the object structure looking for `Balance<SUI>`.

Thanks for the tip on` modified_at_versions`--I will give that a shot. Maybe I actually want to implement this conversation function for effects instead, since it will also make it easier to check conservation in a client if desired.

Reply:
`modified_at_versions` will sometimes include values object keys that don't exist in the store (particularly for objects that were unwrapped in the transaction) -- just something to watch out for.